pre_commit/languages/renv.py

```diff
@@ -0,0 +1,138 @@
+import contextlib
+import os
+import shlex
+import shutil
+from typing import Generator
+from typing import Sequence
+from typing import Tuple
+
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.hook import Hook
+from pre_commit.languages import helpers
+from pre_commit.prefix import Prefix
+from pre_commit.util import clean_path_on_failure
+from pre_commit.util import cmd_output_b
+
+ENVIRONMENT_DIR = 'renv'
+get_default_version = helpers.basic_get_default_version
+healthy = helpers.basic_healthy
+
+
+@contextlib.contextmanager
+def in_env(
+        prefix: Prefix,
+        language_version: str,
+) -> Generator[None, None, None]:
+    envdir = get_env_dir(prefix, language_version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+def get_env_dir(prefix: Prefix, version: str) -> str:
+    return prefix.path(helpers.environment_dir(ENVIRONMENT_DIR, version))
+
+
+def get_env_patch(venv: str) -> PatchesT:
+    return (
+        ('R_PROFILE_USER', os.path.join(venv, 'activate.R')),
+    )
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    env_dir = get_env_dir(prefix, version)
+    with clean_path_on_failure(env_dir):
+        os.makedirs(env_dir, exist_ok=True)
+        shutil.copy(prefix.path('DESCRIPTION'), env_dir)
+        shutil.copy(prefix.path('renv.lock'), env_dir)
+        cmd_output_b(
+            'Rscript', '--vanilla', '-e',
+            """
+            missing_pkgs <- setdiff(
+                "renv", unname(installed.packages()[, "Package"])
+            )
+            options(
+                repos = c(CRAN = "https://cloud.r-project.org"),
+                renv.consent = TRUE
+            )
+            install.packages(missing_pkgs)
+            renv::activate()
+            renv::restore()
+            activate_statement <- paste0(
+              'renv::activate("', file.path(getwd()), '"); '
+            )
+            writeLines(activate_statement, 'activate.R')
+            renv::install(normalizePath('.'))
+            """,
+            cwd=env_dir,
+        )
+        if additional_dependencies:
+            cmd_output_b(
+                'Rscript', '-e',
+                """
+                renv::install(commandArgs(trailingOnly = TRUE))
+                """, *additional_dependencies,
+                cwd=env_dir,
+            )
+
+
+def run_hook(
+        hook: Hook,
+        file_args: Sequence[str],
+        color: bool,
+) -> Tuple[int, bytes]:
+    with in_env(hook.prefix, hook.language_version):
+
+        return helpers.run_xargs(
+            hook, cmd_from_hook(hook), file_args, color=color,
```

Comment:
I usually prefer to read a file from the bottom upwards -- also helper functions that live only in this module should be underscore-prefixed to make it more clear what is/isn't part of the api (the api specification is in the `languages/all.py` file)

Reply:
Ok, I can adapt the order. I can also prefix the helper functions not part of the API. But why is `in_env` and `get_env_patch` not prefixed in other modules although not listed in `languages/all.py`?

Reply:
hmmmm yeah good point, those should probably be underscore-prefixed everywhere -- probably just legacy and copy paste 😅